types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,630 @@
+
+	if (request->invalid) {
+		/* Pre-emptively reject invalid lease requests */
+		wp_drm_lease_v1_send_finished(lease->resource);
+	} else {
+		wlr_signal_emit_safe(&request->device->manager->events.request,
+				request);
+	}
+	wl_resource_destroy(resource);
+}
+
+static struct wp_drm_lease_request_v1_interface lease_request_impl = {
+	.request_connector = drm_lease_request_v1_handle_request_connector,
+	.submit = drm_lease_request_v1_handle_submit,
+};
+
+static void drm_lease_device_v1_handle_resource_destroy(
+		struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+	wl_list_init(wl_resource_get_link(resource));
+}
+
+static void drm_lease_device_v1_handle_stop(
+		struct wl_client *client, struct wl_resource *resource) {
+	wp_drm_lease_device_v1_send_finished(resource);
+	wl_resource_destroy(resource);
+}
+
+static void drm_lease_device_v1_handle_create_lease_request(
+		struct wl_client *client, struct wl_resource *resource, uint32_t id) {
+	struct wlr_drm_lease_device_v1 *device =
+		wlr_drm_lease_device_v1_from_resource(resource);
+
+	struct wlr_drm_lease_request_v1 *req =
+		calloc(1, sizeof(struct wlr_drm_lease_request_v1));
+	if (!req) {
+		wlr_log(WLR_ERROR, "Failed to allocate wlr_drm_lease_request_v1");
+		wl_resource_post_no_memory(resource);
+		return;
+	}
+
+	wlr_log(WLR_DEBUG, "Created request %p", req);
+
+	struct wl_resource *wl_resource = wl_resource_create(client,
+		&wp_drm_lease_request_v1_interface, DRM_LEASE_DEVICE_V1_VERSION, id);
+	if (!wl_resource) {
+		wlr_log(WLR_ERROR, "Failed to allocate wl_resource");
+		wl_resource_post_no_memory(resource);
+		free(req);
+		return;
+	}
+
+	req->device = device;
+	req->resource = wl_resource;
+	req->connectors = NULL;
+	req->n_connectors = 0;
+
+	wl_resource_set_implementation(wl_resource, &lease_request_impl,
+		req, drm_lease_request_v1_handle_resource_destroy);
+
+	wl_list_insert(&device->requests, wl_resource_get_link(wl_resource));
+}
+
+static struct wp_drm_lease_device_v1_interface lease_device_impl = {
+	.stop = drm_lease_device_v1_handle_stop,
+	.create_lease_request = drm_lease_device_v1_handle_create_lease_request,
+};
+
+static void drm_connector_v1_handle_resource_destroy(
+		struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+	wl_list_init(wl_resource_get_link(resource));
+}
+
+static void drm_connector_v1_handle_destroy(
+		struct wl_client *client, struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct wp_drm_lease_connector_v1_interface lease_connector_impl = {
+	.destroy = drm_connector_v1_handle_destroy,
+};
+
+static void drm_lease_connector_v1_send_to_client(
+		struct wlr_drm_lease_connector_v1 *connector,
+		struct wl_client *wl_client, struct wl_resource *device) {
+	if (connector->active_lease) {
+		return;
+	}
+
+	struct wl_resource *wl_resource = wl_resource_create(wl_client,
+			&wp_drm_lease_connector_v1_interface, DRM_LEASE_DEVICE_V1_VERSION,
+			0);
```

Comment:
s/0/NULL/ since this is a `void *` pointer.

Reply:
Isn't the last parameter a `uint32_t`?

```c
struct wl_resource * wl_resource_create(struct wl_client *client, const struct wl_interface *interface, int version, uint32_t id)
```

Reply:
Oh, sorry, mixed that up with `wl_resource_set_implementation`!